crates/containerd-shim-wasm/src/sandbox/shim/otel.rs

```diff
@@ -0,0 +1,164 @@
+///
+/// Returns a `Result` containing the initialized tracer or a `TraceError` if initialization fails.
+///
+/// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#configuration-options
+impl Config {
+    /// Creates a new builder for `OtelConfig`.
+    pub fn builder() -> ConfigBuilder {
+        ConfigBuilder::default()
+    }
+
+    /// Initializes the tracer, sets up the telemetry and subscriber layers, and sets the global subscriber.
+    pub fn init(&self) -> anyhow::Result<ShutdownGuard> {
+        let tracer = self.init_tracer()?;
+        let telemetry = tracing_opentelemetry::layer().with_tracer(tracer);
+        set_text_map_propagator(TraceContextPropagator::new());
+
+        let filter = EnvFilter::try_new("info,h2=off")?;
+
+        let subscriber = Registry::default().with(telemetry).with(filter);
+
+        tracing::subscriber::set_global_default(subscriber)?;
+        Ok(ShutdownGuard)
+    }
+
+    pub fn get_trace_context() -> anyhow::Result<String> {
+        // propogate the context
+        let mut injector: HashMap<String, String> = HashMap::new();
+        global::get_text_map_propagator(|propagator| {
+            // retrieve the context from `tracing`
+            propagator.inject_context(&Span::current().context(), &mut injector);
+        });
+        Ok(serde_json::to_string(&injector)?)
+    }
+
+    pub fn set_trace_context(trace_context: &str) -> anyhow::Result<()> {
+        let extractor: HashMap<String, String> = serde_json::from_str(trace_context)?;
+        let context = global::get_text_map_propagator(|propagator| propagator.extract(&extractor));
+        Span::current().set_parent(context);
+        Ok(())
+    }
+
+    fn init_tracer_http_protobuf(&self) -> SpanExporterBuilder {
+        opentelemetry_otlp::new_exporter()
+            .http()
+            .with_endpoint(&self.otel_endpoint)
+            .into()
+    }
+
+    fn init_tracer_grpc(&self) -> SpanExporterBuilder {
+        opentelemetry_otlp::new_exporter()
+            .tonic()
+            .with_endpoint(&self.otel_endpoint)
+            .into()
+    }
+
+    fn init_tracer(&self) -> Result<opentelemetry_sdk::trace::Tracer, TraceError> {
+        let exporter = match self.otel_protocol.as_str() {
+            OTEL_EXPORTER_OTLP_PROTOCOL_HTTP_PROTOBUF => self.init_tracer_http_protobuf(),
+            OTEL_EXPORTER_OTLP_PROTOCOL_GRPC => self.init_tracer_grpc(),
+            _ => Err(TraceError::from(
+                "Invalid OTEL_EXPORTER_OTLP_PROTOCOL value",
+            ))?,
+        };
+
+        opentelemetry_otlp::new_pipeline()
+            .tracing()
+            .with_exporter(exporter)
+            .with_trace_config(sdktrace::config())
+            .install_batch(runtime::Tokio)
+    }
+}
+
+/// Shutdown of the open telemetry services will automatically called when the OtelConfig instance goes out of scope.
+#[must_use]
+pub struct ShutdownGuard;
```

Comment:
**thought:** It's a bit odd that a drop guard has not members and is guarding a global object. Is there any way we can get a hold of the provider we need to shutdown, instead of relying on the global one?

Reply:
I think this particular SDK usage

```rust
opentelemetry_otlp::new_pipeline()
            .tracing()
            .with_exporter(exporter)
            .with_trace_config(sdktrace::config())
            .install_batch(runtime::Tokio)
```

sets a global tracer provider, and thus the only way to shutdown that is to call `global::shutdown_tracer_provider()`.